File: start.py
import turtle
import tkinter as tk
from Q_learning import (
    QLearningAgent,
)
import random
import pickle


def move_turtle(action):
    x, y = robot.pos()

    if (
        action == "up" and y < 190
    ):  # Slightly less than the boundary to prevent overshooting
        robot.setheading(90)
        robot.forward(10)
    elif action == "down" and y > -190:
        robot.setheading(270)
        robot.forward(10)
    elif action == "left" and x > -190:
        robot.setheading(180)
        robot.forward(10)
    elif action == "right" and x < 190:
        robot.setheading(0)
        robot.forward(10)


def draw_target():
    target.penup()
    target.goto(50, 100)  # Top-left corner of the rectangle
    target.pendown()
    for _ in range(2):
        target.forward(40)  # Width of the rectangle
        target.right(90)
        target.forward(80)  # Height of the rectangle
        target.right(90)
    target.penup()


def check_target():
    x, y = robot.pos()
    if 50 <= x <= 90 and 20 <= y <= 100:
        return True
    return False


colors = ["red", "green", "blue", "yellow", "purple", "orange"]


def train_robot(agent, step_limit=200):
    success_count = 0  # Counter for successful episodes

    for episode in range(1000):
        state = (0, 0)  # Starting position
        robot.goto(0, 0)
        color = random.choice(colors)
        robot.pencolor(color)

        for step in range(step_limit):
            action = agent.choose_action(state)
            move_turtle(action)

            new_state = (int(robot.xcor()), int(robot.ycor()))
            print(
                f"Episode: {episode}, Successes: {success_count}, Step: {step}, State: {state}, Action: {action}, New State: {new_state}"
            )

            if check_target():
                reward = 100  # Positive reward for reaching the target
                agent.update_q_table(state, new_state, action, reward)
                success_count += 1
                print(
                    f"Success! Episode: {episode}, Step: {step}, Total Successes: {success_count}"
                )
                break
            else:
                reward = -1  # Negative reward for each step

            # Update the Q-table
            agent.update_q_table(state, new_state, action, reward)
            state = new_state

            if step == step_limit - 1:
                reward = (
                    -100
                )  # Additional punishment for not reaching the target in time
                agent.update_q_table(state, new_state, action, reward)
                robot.goto(0, 0)  # Reset turtle's position
                state = (0, 0)  # Also reset the state

        # Update exploration rate in the agent
        agent.exploration_rate = max(
            agent.min_exploration_rate,
            agent.exploration_rate * agent.exploration_decay_rate,
        )

    # Save the Q-table to a file after training is complete
    with open("q_table.pkl", "wb") as f:
        pickle.dump(agent.q_table, f)

    print(f"Training complete with {success_count} successes.")
    window.textinput(
        "Training Complete",
        "The training process has finished. Total successes: " + str(success_count),
    )


# Setting up the screen and turtle
window = turtle.Screen()
window.title("TurtleSim Robot Simulation")
window.screensize(canvwidth=400, canvheight=400)  # Adjust the drawing area size
window.setup(width=500, height=500)

robot = turtle.Turtle()
robot.shape("turtle")

target = turtle.Turtle()
target.color("blue")
target.hideturtle()
draw_target()

# Initialize QLearningAgent and start training
agent = QLearningAgent()
train_robot(agent)

window.mainloop()

File: cpp/start.cpp
#include <iostream>
#include <vector>
#include <string>
#include <fstream>
#include <nlohmann/json.hpp>
#include "Q_learning.cpp"

// Function to move the robot in the simulated environment
std::pair<int, int> move_robot(const std::pair<int, int> &state, const std::string &action)
{
    int x = state.first, y = state.second;
    if (action == "up" && y < 190)
        y += 10;
    else if (action == "down" && y > -190)
        y -= 10;
    else if (action == "left" && x > -190)
        x -= 10;
    else if (action == "right" && x < 190)
        x += 10;
    return {x, y};
}

// Function to check if the robot reached the target
bool check_target(const std::pair<int, int> &state)
{
    return (state.first >= 50 && state.first <= 90 && state.second >= 20 && state.second <= 100);
}

// Training function
void train_robot(QLearningAgent &agent, int step_limit = 200)
{
    int success_count = 0;

    for (int episode = 0; episode < 1000; ++episode)
    {
        std::pair<int, int> state = {0, 0}; // Starting position
        int total_reward_for_episode = 0;

        for (int step = 0; step < step_limit; ++step)
        {
            std::string action = agent.choose_action(state);
            std::pair<int, int> new_state = move_robot(state, action);
            int reward = -1; // Negative reward for each step

            if (check_target(new_state))
            {
                reward = 100;
                success_count++;
                break;
            }

            total_reward_for_episode += reward;
            agent.update_q_table(state, new_state, action, reward);
            state = new_state;

            if (step == step_limit - 1)
            {
                // Additional punishment for not reaching the target in time
                agent.update_q_table(state, new_state, action, -100);
            }
        }

        // Update exploration rate in the agent
        agent.update_exploration_rate(episode);
    }

    std::cout << "Training complete with " << success_count << " successes." << std::endl;

    // Save the Q-table to a file
    std::ofstream file("q_table.json");
    nlohmann::json j = agent.get_q_table(); // Assuming QLearningAgent has a method to return Q-table in JSON format
    file << j.dump();
    file.close();
}

int main()
{
    QLearningAgent agent;
    train_robot(agent);
    return 0;
}
